Extension cases

kaapstorm opened the pull request:
Extension cases progress so far. I still have four tests to write: Updating and preloading extension cases from a host case in basic and advanced modules. (They are the tests that are currently skipped with a "TODO".)

Sorry this is such a big chunk all at once. Might be easier to go through individual commits. Let me know if you think I should rebase and squash some of these commits to make them easier to follow.

@snopoke, @orangejenny, @biyeun

kaapstorm committed:
Added test stubs

kaapstorm committed:
Add relationship index attribute

kaapstorm committed:
A little refactor

kaapstorm committed:
Validate index relationship

kaapstorm committed:
Tests and test stubs

kaapstorm committed:
Validate action relationship

kaapstorm committed:
Add toggle

kaapstorm committed:
Rename child case to subcase to avoid confusion

kaapstorm committed:
Add relationship to case management

kaapstorm committed:
Set relationship choices on CommCareCaseIndex

kaapstorm committed:
Drop underscore

kaapstorm committed:
More tests

kaapstorm committed:
Rename hostcase to host

kaapstorm committed:
Added CaseBlock.add_index_ref test

kaapstorm committed:
Test updates

kaapstorm committed:
Remove basic case management changes (for now?)

kaapstorm committed:
Add relationship to advanced app builder

kaapstorm committed:
Migration and lazy migration for parent indices

kaapstorm committed:
Updated references to parent_tag and parent_reference_id

(Tests still to be updated)

kaapstorm committed:
Replaced parent_tag with parents in advanced app builder classes

kaapstorm committed:
Checked build errors

kaapstorm committed:
Make parentIndex an observableArray

kaapstorm committed:
UI adds parents to an array

kaapstorm committed:
parent_reference_id to parent.reference_id

kaapstorm committed:
Preserve existing behaviour

Only render relationship attribute for non-child relationship.

kaapstorm committed:
A little cleaning up

kaapstorm committed:
Added AdvancedAction migration tests

kaapstorm committed:
Updated app_manager tests to use ParentIndex

kaapstorm committed:
Add #host and #parent to form/module filters

kaapstorm committed:
Use CaseIDXPath to get parent and host cases

kaapstorm committed:
Added host update and preload tests

dimagimon commented:
There are 7 errors:

* corehq/apps/app_manager/suite_xml.py, line 1902 - E501 line too long (117 > 115 characters)
* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 712 - Missing semicolon.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 928 - Missing semicolon.

kaapstorm committed:
Fix tests, appease dimagimon

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent

snopoke commented on corehq/apps/app_manager/management/commands/migrate_adv_form_action_parent.py:
```diff
@@ -0,0 +1,33 @@
+                for action_name in form.get('actions', {}):
+                    if action_name in ('load_update_cases', 'open_cases'):
+                        for action in form['actions'][action_name]:
+                            if 'parent_tag' in action:
```
Could be `if action.get('parent_tag', None)` to reduce the nesting by 1.

kaapstorm replied:
if `action['parent_tag']` is `None` or `''` I still want to remove the unused keys, and set `should_save = True` ... although I'm assuming I need to clean up `action`. Is that assumption right? Do I need to worry about that, or can I just leave `action` as-is?

snopoke replied:
👍

On 5 August 2015 at 10:29, Norman Hooper <notifications@github.com> wrote:

> In
> corehq/apps/app_manager/management/commands/migrate_adv_form_action_parent.py
> <https://github.com/dimagi/commcare-hq/pull/7749#discussion_r36279492>:
>
> > +
> > +
> > +class Command(AppMigrationCommandBase):
> > +    help = "Migrate single parent index to ParentIndex list in advanced form actions."
> > +
> > +    include_builds = False  # AdvancedAction lazy-migrates reverted builds
> > +
> > +    def migrate_app(self, app_doc):
> > +        modules = [m for m in app_doc['modules'] if m.get('module_type', '') == 'advanced']
> > +        should_save = False
> > +        for module in modules:
> > +            for form in module['forms']:
> > +                for action_name in form.get('actions', {}):
> > +                    if action_name in ('load_update_cases', 'open_cases'):
> > +                        for action in form['actions'][action_name]:
> > +                            if 'parent_tag' in action:
>
> if action['parent_tag'] is None or '' I still want to remove the unused
> keys, and set should_save = True ... although I'm assuming I need to
> clean up action. Is that assumption right? Do I need to worry about that,
> or can I just leave action as-is?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7749/files#r36279492>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -329,12 +334,17 @@ def all_property_names(self):
         return names
 
 
+class ParentIndex(DocumentSchema):
```
Maybe call this `CaseIndex`?

snopoke replied:
I see it does tie in with `action.parents` but just wondering if its worth moving away from 'parent' metaphor in the models and sticking closer to language of the XML spec i.e. index / indices

snopoke replied:
Probably not worth it at this point though since it would be a lot of changes.

kaapstorm replied:
I agree with all of that. I should have gone with `action.indices` from the beginning. :expressionless:  Oh well.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -352,7 +362,28 @@ def get_property_names(self):
+                else:
+                    data['parents'] = [parent]
+                del data['parent_tag']
+                data.pop('parent_reference_id', None)
```
Why use `del` above and `pop` here?

kaapstorm replied:
We can use `del` because we know the key exists. If `parent_reference_id` isn't in `data`, `del` will throw a KeyError. `somedict.pop(somekey, None)` won't, but it will do the same thing as `del`. It's more wordy, so I like `del` more.

kaapstorm replied:
Actually, do we need to bother removing unused keys from `data`? Or is it better/safer to clean up?

snopoke replied:
Its good to clean up else next time it will still be there.

On 5 August 2015 at 10:24, Norman Hooper <notifications@github.com> wrote:

> In corehq/apps/app_manager/models.py
> <https://github.com/dimagi/commcare-hq/pull/7749#discussion_r36279166>:
>
> > +        _assert = soft_assert(to=['nhooper@dimagi.com'])
> > +        if not _assert('parent_tag' not in data,
> > +                       'AdvancedAction not migrated. Old application build may have been reverted.'):
> > +            if data['parent_tag']:
> > +                # Lazy-migrate single parent index to list of parent/host indices
> > +                parent = {
> > +                    'tag': data['parent_tag'],
> > +                    'reference_id': data.get('parent_reference_id', 'parent'),
> > +                    'relationship': data.get('relationship', 'child')
> > +                }
> > +                if hasattr(data.get('parents'), 'append'):
> > +                    data['parents'].append(parent)
> > +                else:
> > +                    data['parents'] = [parent]
> > +                del data['parent_tag']
> > +                data.pop('parent_reference_id', None)
>
> Actually, do we need to bother removing unused keys from data? Or is it
> better/safer to clean up?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7749/files#r36279166>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2291,10 +2334,12 @@ def add_insert_form(self, from_module, form, index=None, with_source=False):
                             details_module=module.unique_id,
-                            parent_tag='_'.join(['parent'] * (i + 2)) if n > 0 else ''
+                            parents=[
+                                ParentIndex(tag='_'.join(['parent'] * (i + 2)) if n > 0 else '')
```
Since you've added the `relationship` field to basic modules you should probably  copy that over as well.

snopoke replied:
Also below.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -4467,6 +4514,14 @@ def get_hierarchy(case_type):
+        """
+        return {i for m in self.get_modules()
+                if m.case_type == module.case_type
+                for i in m.get_subcase_types()}
```
`i` is a misleading variable name since it usually refers to a loop iteration count.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -4467,6 +4514,14 @@ def get_hierarchy(case_type):
 
         return meta
 
+    def get_subcase_types_of_module(self, module):
```
Could make this more generic by just passing the case type rather than the module.

snopoke commented on corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js:
```diff
@@ -419,6 +424,21 @@ var AdvancedCase = (function () {
+        self.reference_id = ko.observable('parent');
+        self.relationship = ko.observable('child');
+        self.relationship.subscribe(function (rel) {
+            if (rel == 'extension' && self.reference_id() == 'parent') {
```
Should use `===`. Also no real need for the second check, you could always set the reference_id even if its already set to the correct value.

kaapstorm replied:
My thinking with the second check was that if the user set a reference ID (the field appears before the relationship field), and then set the relationship, and HQ changed the reference ID they'd just set, that would be annoying.

snopoke commented on corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js:
```diff
@@ -527,6 +552,35 @@ var AdvancedCase = (function () {
+            var blankParent = new ParentIndex();
+            self.newParent.tag(blankParent.tag());
+            self.newParent.reference_id(blankParent.reference_id());
+            self.newParent.relationship(blankParent.relationship());
```
Can you not just assign `new ParentIndex()` to `self.newParent`?

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1907,33 +1922,26 @@ def get_target_module(case_type, module_id, with_product_details=False):
-                        parent_action.case_session_var
-                    )
+                if action.parents:
+                    for parent in action.parents:
```
I don't think this is correct. This would load the same case multiple times by using different parent references. Although a case may have multiple indices you can only load it into the session based off one of those and not both.

So we should add some validation (not here) to make sure that there is only one parent for any `load` actions. Alternatively we should split the model so that `load` actions have a single parent and `open` actions can have multiple.

kaapstorm commented on corehq/apps/app_manager/xform.py:
```diff
@@ -461,12 +461,22 @@ def add_close_block(self, relevance):
+        if relationship not in ('child', 'extension'):
+            raise CaseError('Valid values for an index relationship are "child" and "extension"')
+        if relationship == 'child':
+            parent_index = make_case_elem(reference_id, {'case_type': case_type})
```
(Leave out `relationship` attribute to preserve current behaviour.)

kaapstorm committed:
Remove extension cases feature flag

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent

kaapstorm committed:
More generic method, slightly better var name

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent

kaapstorm committed:
ParentIndex renamed to CaseIndex

(git grep is my friend. PyCharm's refactor doesn't catch all JavaScript
scenarios.)

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent

kaapstorm committed:
Use JavaScript "==="

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent

kaapstorm committed:
Refactor action.parents to action.case_indices

Also includes validation for LoadUpdateAction case_index

dimagimon commented:
There are 12 errors:

* corehq/apps/app_manager/models.py, line 5 - F401 'soft_assert' imported but unused
* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 311 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 312 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 313 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 315 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 316 - 'action' used out of scope.

kaapstorm committed:
Removed trailing comma

dimagimon commented:
There are 12 errors:

* corehq/apps/app_manager/models.py, line 5 - F401 'soft_assert' imported but unused
* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 311 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 312 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 313 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 315 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 316 - 'action' used out of scope.

kaapstorm commented:
Refactored AdvancedAction.parents to LoadUpdateAction.case_index and AdvancedOpenCaseAction.case_indices. Advanced app builder treats load/update case *parent* and open case *parents* correctly.

kaapstorm committed:
Missed a "parents".

dimagimon commented:
There are 12 errors:

* corehq/apps/app_manager/models.py, line 5 - F401 'soft_assert' imported but unused
* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 311 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 312 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 313 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 315 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 316 - 'action' used out of scope.

kaapstorm committed:
caseIndex.tag is an observable

dimagimon commented:
There are 12 errors:

* corehq/apps/app_manager/models.py, line 5 - F401 'soft_assert' imported but unused
* corehq/apps/app_manager/tests/test_extension_case.py, line 162 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 163 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 188 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 189 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 214 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/tests/test_extension_case.py, line 215 - E131 continuation line unaligned for hanging indent
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 311 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 312 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 313 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 315 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 316 - 'action' used out of scope.

kaapstorm committed:
Move case.mock.CaseBlock tests to case.tests

dimagimon commented:
There are 13 errors:

* corehq/apps/app_manager/models.py, line 5 - F401 'soft_assert' imported but unused
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 191 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 192 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 217 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 218 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 243 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 244 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 267 - W391 blank line at end of file
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 311 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 312 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 313 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 315 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 316 - 'action' used out of scope.

kaapstorm committed:
self.case_index.tag is an observable

kaapstorm committed:
Remove subcase checkbox from open case action

kaapstorm committed:
Spell "ID" the same everywhere

kaapstorm committed:
Fix UI

dimagimon commented:
There are 13 errors:

* corehq/apps/app_manager/models.py, line 5 - F401 'soft_assert' imported but unused
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 191 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 192 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 217 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 218 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 243 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 244 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 267 - W391 blank line at end of file
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 310 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 311 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 312 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 314 - 'action' used out of scope.
* corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js, line 315 - 'action' used out of scope.

kaapstorm committed:
Feeding the dimagimon(ster)

dimagimon commented:
There are 6 errors:

* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 191 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 192 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 217 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 218 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 243 - E131 continuation line unaligned for hanging indent
* corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py, line 244 - E131 continuation line unaligned for hanging indent

kaapstorm committed:
The things we do for dimagimon

kaapstorm commented:
\o/ All done ... \<cough>except those four still-missing tests\</cough>

kaapstorm committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

kaapstorm committed:
Mock is_usercase_in_use

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -393,6 +402,25 @@ class LoadUpdateAction(AdvancedAction):
     auto_select = SchemaProperty(AutoSelectCase, default=None)
     show_product_stock = BooleanProperty(default=False)
     product_program = StringProperty()
+    case_index = SchemaProperty(CaseIndex)
```
Not sure if you want to do this or not but if you add `default=None` then JsonObject won't create an object if there isn't one.

kaapstorm replied:
Thanks. I think that's probably the right thing to do, instead of checking the value of tag.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -4572,6 +4646,14 @@ def get_hierarchy(case_type):
+        """
+        return {t for m in self.get_modules()
+                if m.case_type == case_type
+                for t in m.get_subcase_types()}
```
This won't necessarily work correctly since Advanced forms can have different case types to the module.

kaapstorm replied:
Hmmm. This originally came out of views.py (https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/views.py#L797-L801). The behaviour should be unchanged, but you make a good point!

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -393,6 +402,25 @@ class LoadUpdateAction(AdvancedAction):
+
+    @property
+    def case_indices(self):
+        # Allows us to ducktype AdvancedOpenCaseAction
```
Nice

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1950,33 +1965,25 @@ def get_target_module(case_type, module_id, with_product_details=False):
+                        'case_type': action.case_type,
+                        'requires_selection': True,
+                        'action': action
+                    })
```
This change doesn't seem like an improvement. I think originally you changed it since you were looping through the indices but now there is just one and so we can revert to the previous logic where the if/else just determines what the parent_filter value should be.

snopoke commented on corehq/apps/app_manager/templates/app_manager/partials/case_config_advanced.html:
```diff
@@ -79,23 +87,38 @@
+                                       value: relationship"></select>
+                    <br />
+
+                    <button data-bind="click: $parent.addCaseIndex">Add</button>
```
Would it not be easier to just have an 'add' button that pushes a new CaseIndex into the list rather than having to copy values from the 'newCaseIndex' into the list each time?

kaapstorm replied:
Thanks. I'll check that out. I couldn't push it and make a new instance because I bind newCaseIndex (https://github.com/dimagi/commcare-hq/blob/extension_cases/corehq/apps/app_manager/templates/app_manager/partials/case_config_advanced.html#L103-103). If I don't bind it, it might work.

snopoke replied:
What I mean is instead of having this here at all just have a button that
pushes an empty index into the list which should make it appear in the list
above.

On 7 August 2015 at 11:31, Norman Hooper <notifications@github.com> wrote:

> In
> corehq/apps/app_manager/templates/app_manager/partials/case_config_advanced.html
> <https://github.com/dimagi/commcare-hq/pull/7749#discussion_r36504076>:
>
> >                      {% endif %}
> > +
> > +                    <label>{% trans "Relationship:" %}</label>
> > +                    <select data-bind="options: $parent.relationshipTypes,
> > +                                       optionsCaption: 'Child case or extension case?',
> > +                                       value: relationship"></select>
> > +                    <br />
> > +
> > +                    <button data-bind="click: $parent.addCaseIndex">Add</button>
>
> Thanks. I'll check that out. I couldn't push it and make a new instance
> because I bind newCaseIndex (
> https://github.com/dimagi/commcare-hq/blob/extension_cases/corehq/apps/app_manager/templates/app_manager/partials/case_config_advanced.html#L103-103).
> If I don't bind it, it might work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7749/files#r36504076>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

snopoke commented on corehq/apps/app_manager/tests/data/extension_case/preload_host_case.xml:
```diff
@@ -0,0 +1,52 @@
+      <bind calculate="/data/meta/userID" nodeset="/data/case/@user_id"/>
+      <bind calculate="instance('commcaresession')/session/data/case_id" nodeset="/data/case/@case_id"/>
+      <setvalue event="xforms-ready" ref="/data/question1" value="instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]/question1"/>
+      <setvalue event="xforms-ready" ref="/data/question2" value="instance('casedb')/casedb/case[@case_id=instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]/index/host]/question2"/>
```
There is not `/data/question2` in this form

kaapstorm replied:
You have the eyes of an eagle!

benrudolph replied:
:bird:

snopoke commented on corehq/apps/app_manager/tests/data/extension_case/preload_host_case_adv.xml:
```diff
@@ -0,0 +1,48 @@
+        </translation>
+      </itext>
+      <setvalue event="xforms-ready" ref="/data/question1" value="instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_load_1]/question1"/>
+      <setvalue event="xforms-ready" ref="/data/question2" value="instance('casedb')/casedb/case[@case_id=instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_load_1]/index/host]/question2"/>
```
again, no `/data/question2`

snopoke commented:
looking good

kaapstorm committed:
Simplify suite_xml

kaapstorm committed:
Add the phantom question2

snopoke commented:
Test fail real:
```
======================================================================
FAIL: test_xform_case_block_index_default_relationship (corehq.apps.app_manager.tests.test_extension_case.CaseBlockIndexRelationshipTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/test_extension_case.py", line 201, in test_xform_case_block_index_default_relationship
    self.assertXmlEqual(self.get_xml('open_subcase_child'), str(self.xform))
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/util.py", line 26, in assertXmlEqual
    _check_shared(expected, actual, LXMLOutputChecker(), "xml")
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/util.py", line 113, in _check_shared
    raise AssertionError(message)
AssertionError: XML mismatch
--- want.xml
+++ got.xml
@@ -5,6 +5,7 @@
       <instance>
         <data xmlns:jrm="http://dev.commcarehq.org/jr/xforms" xmlns="http://openrosa.org/formdesigner/A22A5D53-037A-48DE-979B-BAA54734194E" name="New Form" uiVersion="1" version="1">
           <question1/>
+          <question2/>
           <subcase_0>
             <case xmlns="http://commcarehq.org/case/transaction/v2" case_id="" date_modified="" user_id="">
               <create>
@@ -23,10 +24,14 @@
         </data>
       </instance>
       <bind nodeset="/data/question1" type="xsd:string"/>
+      <bind nodeset="/data/question2" type="xsd:string"/>
       <itext>
         <translation default="" lang="en">
           <text id="question1-label">
             <value>question1</value>
+          </text>
+          <text id="question2-label">
+            <value>question2</value>
           </text>
         </translation>
       </itext>
@@ -44,5 +49,8 @@
     <input ref="/data/question1">
       <label ref="jr:itext('question1-label')"/>
     </input>
+    <input ref="/data/question2">
+      <label ref="jr:itext('question2-label')"/>
+    </input>
   </h:body>
 </h:html>
```

kaapstorm committed:
Missed two question2s

kaapstorm committed:
Drop newCaseIndex

And remember to translate labels

kaapstorm commented:
@snopoke the array of case indices now does what I think you're suggesting: you add a case index to the array, and then you can set the parent tag, etc.

I also tried setting case_index default=None, and the Python stuff was easy to refactor, but the Knockout stuff is not behaving the way I expect. I've stashed the changes for now.

snopoke commented:
👍 I'll check it out

On 7 August 2015 at 16:39, Norman Hooper <notifications@github.com> wrote:

> @snopoke <https://github.com/snopoke> the array of case indices now does
> what I think you're suggesting: you add a case index to the array, and then
> you can set the parent tag, etc.
>
> I also tried setting case_index default=None, and the Python stuff was
> easy to refactor, but the Knockout stuff is not behaving the way I expect.
> I've stashed the changes for now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7749#issuecomment-128716826>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

kaapstorm committed:
Hard-code date_modified

kaapstorm committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/ex-submodules/casexml/apps/case/tests/test_indexes.py

snopoke commented:
@kaapstorm looks good. A few UI comments though:

* it would be nice if we could get all the form elements on the same line rather than taking up 3 lines
* can we style the 'add parent index' button
* I think the text for the button should be 'add case index'

For styling I think trying to follow what the case list looks like would be good (except when there are no indexes we can hide the table completely and just show the button).

dannyroberts committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

Easy merge, just imports

kaapstorm committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

kaapstorm committed:
Merge branch 'extension_cases' of github.com:dimagi/commcare-hq into extension_cases

kaapstorm committed:
Circular ref build error no longer applicable

kaapstorm committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

dimagimon commented on corehq/apps/app_manager/tests/test_suite.py:
```diff
@@ -641,6 +666,392 @@ def test_report_module(self):
+        self._load_case(m1f0, 'gold-fish')
+        self._load_case(m1f0, 'guppy', parent_module=self.module_0)
+
+        self.assertXmlPartialEqual(self.get_xml('child-module-entry-datums-added'), self.app.create_suite(), "./entry")
```
E501 line too long (119 > 115 characters)

orangejenny committed:
merge master

kaapstorm committed:
Merge branch 'master' into extension_cases

kaapstorm committed:
Fix test for parent case handling

Not so sure about the detail-confirm of the parent case, but going with this for now.

kaapstorm committed:
parent_tag -> case_indices

kaapstorm committed:
Merge branch 'master' into extension_cases

kaapstorm committed:
Revert "Fix test for parent case handling"

This reverts commit 337ca181c6264f9c27fda7a1cbcd5c40903a9636.

kaapstorm committed:
get_open_subcase_actions should not return dups

kaapstorm committed:
Replace parent_tag in tests

kaapstorm committed:
Use set comprehension

kaapstorm committed:
form.session_var_for_action only takes one param

kaapstorm committed:
Fixed test_suite.py merge

kaapstorm committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/apps/app_manager/tests/test_models.py

kaapstorm committed:
OCD

kaapstorm committed:
Fix tests

kaapstorm committed:
Should assertTrue

kaapstorm commented:
Tests passing. Still outstanding: 
* UI tweaks @snopoke suggested 
* Ability to load and update extension case properties from Case Management the way one can with parent/host properties

@snopoke, do you think we should merge now and address those later, or should we address those first?

kaapstorm committed:
Merge branch 'master' into extension_cases

Conflicts:
	corehq/apps/app_manager/tests/test_child_module.py

kaapstorm commented:
Merge conflict resolved

sheelio commented:
@kaapstorm Are there any open items for this?  Would be good to merge and deploy to India - we'd like to use this in a couple of forms.

kaapstorm commented:
@sheelio, I think Simon intends to merge today, and was waiting for the merge conflict to be resolved. There are two items still to address (UI improvements, and ability to modify an extension case from a host case the way you can modify a parent from a child using "parent/"), but I'll do that in a different branch after this has been merged.

sheelio commented:
@kaapstorm 

Access extension case from host  - we don't need this for ICDS.  Might also be a tougher to do since a host case can have multiple extensions.  We've been using model iteration to manage this in the form builder.

fyniac commented:
@kaapstorm an you please make sure it's merged today?  Will have an India deploy happen once this is in

czue commented:
@snopoke should you be point on final review? or @kaapstorm who should be?

dannyroberts commented:
@czue we agreed I'm reviewing it, which I'm doing now

kaapstorm commented:
@czue, I believe that's @dannyroberts

emord commented:
Do I need to run the managment command after I deploy?

kaapstorm commented:
@emord, you should, but it'll work if you don't.

czue commented:
:boom: :fireworks:
